aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import java.io.IOException;
+import java.io.Serializable;
+import org.apache.http.HttpStatus;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.util.SerializableSupplier;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.Delete;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+/**
+ * FileIO implementation backed by S3.  Locations used must follow the conventions for URIs (e.g. s3://bucket/path/..).
+ */
+public class S3FileIO implements FileIO, Serializable {
+  private final SerializableSupplier<S3Client> s3;
+  private transient S3Client client;
+
+  public S3FileIO(SerializableSupplier<S3Client> s3) {
+    this.s3 = s3;
+  }
+
+  @Override
+  public InputFile newInputFile(String path) {
+    return new S3InputFile(new AmazonS3URI(path));
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return new S3OutputFile(new AmazonS3URI(path));
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    AmazonS3URI location = new AmazonS3URI(path);
+    ObjectIdentifier objectIdentifier = ObjectIdentifier.builder().key(location.getKey()).build();
+    Delete delete = Delete.builder().objects(objectIdentifier).build();
+    DeleteObjectsRequest deleteRequest =
+        DeleteObjectsRequest.builder().bucket(location.getBucket()).delete(delete).build();
+
+    client().deleteObjects(deleteRequest);
+  }
+
+  private S3Client client() {
+    if (client == null) {
+      client = s3.get();
+    }
+    return client;
+  }
+
+  public class S3InputFile implements InputFile {
```

Comment:
Is it better to not nest this class inside `S3FileIO`? In this way it is more consistent with `HadoopInputFile`

Reply:
And same comment for `S3OutputFile`

Reply:
I don't think there's any need to be consistent across these implementations.  Also in this case they share the client from S3FileIO, which would lend itself an inner class implementation.  I feel like this is more a stylistic difference and with how little code is involved in the inner classes, I find it more cohesive.

Reply:
Ok, so after saying that, I did end up splitting it out due to implementing `create` causing me to rethink some of the shared functionality between input/output.  It got a bit more complicated and was probably more logic than should be embedded as inner classes.  They're now split out.